File: ipc-codegen/echo_example/cpp/src/echo_server.cpp
// Echo IPC server (C++) over the header-only generated dispatch and
// ipc-runtime's socket/shared-memory server.
// Usage: echo_server --socket /tmp/echo.sock

#include "echo_handlers.hpp"
#include "generated/echo_ipc_server.hpp"

#include <iostream>
#include <string_view>

int main(int argc, char **argv) {
  const char *socket_path = nullptr;
  for (int i = 1; i < argc - 1; i++) {
    if (std::string_view(argv[i]) == "--socket")
      socket_path = argv[i + 1];
  }
  if (!socket_path) {
    std::cerr << "Usage: echo_server --socket <path>\n";
    return 1;
  }

  echo::EchoCtx ctx;
  echo::serve(socket_path, ctx);
  return 0;
}

File: ipc-codegen/echo_example/rust/src/echo_server.rs
//! Echo IPC server — uses GENERATED dispatch + types + ipc-runtime transport.
//! Usage: echo_server --socket /tmp/echo.sock

use echo_wire_compat::handler::EchoHandler;
use ipc_runtime::IpcServer;
use std::cell::RefCell;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let socket_path = args
        .iter()
        .position(|a| a == "--socket")
        .and_then(|i| args.get(i + 1))
        .expect("Usage: echo_server --socket <path>");

    let _ = std::fs::remove_file(socket_path);

    // Wrap handler in RefCell so the FnMut closure can borrow mutably across
    // dispatches.
    let handler = RefCell::new(EchoHandler);

    let mut server = IpcServer::from_path(socket_path).expect("IpcServer::from_path");
    server.install_default_signal_handlers();
    server.listen().expect("IpcServer::listen");

    server.run(|_client_id, payload| {
        echo_wire_compat::generated::echo_server::handle_request(
            &mut *handler.borrow_mut(),
            payload,
        )
    });
}
